src/fn_call.rs

```diff
@@ -71,8 +76,195 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
+            },
+            "__rust_start_panic" => {
+                // This function has the signature:
+                // 'fn __rust_start_panic(payload: usize) -> u32;'
+                //
+                // The caller constructs 'payload' as follows
+                // 1. We start with a type implementing core::panic::BoxMeUp
+                // 2. We make this type into a trait object, obtaining a '&mut dyn BoxMeUp'
+                // 3. We obtain a raw pointer to the above mutable reference: that is, we make:
+                //    '*mut &mut dyn BoxMeUp'
+                // 4. We convert the raw pointer to a 'usize'
+                //
+
+                // When a panic occurs, we (carefully!) reverse the above steps
+                // to get back to the actual panic payload
+                //
+                // Even though our argument is a 'usize', Miri will have kept track
+                // of the fact that it was created via a cast from a pointer.
+                // This allows us to construct an ImmTy with the proper layout,
+                // and dereference it
+                //
+                // Reference:
+                // https://github.com/rust-lang/rust/blob/9ebf47851a357faa4cd97f4b1dc7835f6376e639/src/libpanic_unwind/lib.rs#L101
+                // https://github.com/rust-lang/rust/blob/9ebf47851a357faa4cd97f4b1dc7835f6376e639/src/libpanic_unwind/lib.rs#L81
+                //
+                // payload_raw now represents our '&mut dyn BoxMeUp' - a fat pointer
+                //
+                // Note that we intentially call deref_operand before checking
+                // This ensures that we always check the validity of the argument,
+                // even if we don't end up using it
+               
+                trace!("__rustc_start_panic: {:?}", this.frame().span);
+
+
+                // Read our 'usize' payload argument (which was made by casting
+                // a '*mut &mut dyn BoxMeUp'
+                let payload_raw = this.read_scalar(args[0])?.not_undef()?;
+
+                // Construct an ImmTy, using the precomputed layout of '*mut &mut dyn BoxMeUp'
+                let imm_ty = ImmTy::from_scalar(
+                    payload_raw,
+                    this.machine.cached_data.as_ref().unwrap().box_me_up_layout
+                );
+
+                // Convert our ImmTy to an MPlace, and read it
+                let mplace = this.ref_to_mplace(imm_ty)?;
+
+                // This is an '&mut dyn BoxMeUp'
+                let payload_dyn = this.read_immediate(mplace.into())?;
+
+                // We deliberately do this after we do some validation of the
+                // 'payload'. This should help catch some basic errors in
+                // the caller of this function, even in abort mode
+                if this.tcx.tcx.sess.panic_strategy() == PanicStrategy::Abort  {
+                    return err!(MachineError("the evaluated program abort-panicked".to_string()));
```

Comment:
Add a fixme so in the future we'll debug format the payload and print it.

Reply:
Added